### PR TITLE
Migrate to Tera

### LINKED ISSUE #884: CrateDetails should not be optional
The more I look at this, the more I think this `Option` is not correct. If it's `None`, that means that no crate exists for the (name, version) pair: https://github.com/rust-lang/docs.rs/blob/6a53a71d2ceb032359edc06967099847ecb934d0/src/web/crate_details.rs#L107
So we should either `unwrap()` because we've guaranteed that this crate exists, or return a 404. I'm ok with making this a follow-up PR. The good news is this means the templates never have to handle the case when `details` is None.

_Originally posted by @jyn514 in https://github.com/rust-lang/docs.rs/pull/879_

### PR DESCRIPTION
**Pull request by Kixiron**
Fixup of #879

* Fixed ISEs from rustdoc & details page, tera errors on OOB indexes while handlebars doesn't
* Highlighting was corrected, rustdoc provides its own [link](https://github.com/rust-lang/docs.rs/pull/885#issuecomment-655156459)
* All builds are not shown as failed [link](https://github.com/rust-lang/docs.rs/pull/885#issuecomment-655149288)
* Fixed issue with rendering crate pages: `details.rustdoc` must be piped to `safe` [link](https://github.com/rust-lang/docs.rs/pull/879#discussion_r449815745)

<details>
<summary>Fixed page</summary>

image

</details>

* Added `cexpect!` to getting the crate details, closing #884

Closes #884

### CODE REVIEW

File: templates/rustdoc/page.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                    window.location.hash = hash;
+                    doc_body.focus();
+                }, 1);
+            }
+        } else {
+            doc_body.focus();
+        }
+    </script>
+
+</html>
```
Comment (jyn514) [inline templates/rustdoc/page.html:50]: Missing newline at EOF
```suggestion
</html>
```

**Comment by Kixiron** [comment]
None are possible to be tested without full builds and full ui testing. Ui testing will be a follow up, it’s been blocked on this for a while

**Comment by pietroalbini** [comment]
We can definitely test them without full UI tests: just use kuchiki to parse the HTML returned by the web server, [like I did in this test](https://github.com/rust-lang/docs.rs/blob/7903abc630a0d444f6a2d2f8b4c6991a543ae86e/src/web/releases.rs#L1124-L1166).

**Comment by jyn514** [comment]
Marking as P-high since this is blocking https://github.com/rust-lang/docs.rs/pull/873.

**Comment by jyn514** [comment]
Everything but highlight.js now has tests (since I'm not really sure how to test CSS without full UI tests).

**Comment by Kixiron** [comment]
I don't think the highlighting is a *massive* deal for now, it's not a breakage so much as an annoyance (That is, it doesn't break anything or hinder site usage, it just doesn't look as good)

**Comment by jyn514** [comment]
Yeah, I'm fine merging this as-is if you've tested locally it's fixed now.

**Comment by Kixiron** [comment]
Tested locally and we're all good

File: templates/macros.html [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                <a href="{{ release_url }}" class="pure-menu-link warn" title="{{ release_name }} is yanked">
+                    <i class="fa fa-fw fa-warning"></i> {{ release.version }}
+                </a>
+
+            {# If the release is yanked but built, display an warning #}
+            {%- elif release.yanked and not release.build_status -%}
+                <a href="{{ release_url }}" class="pure-menu-link warn"
+                    title="{{ release_name }} is yanked and docs.rs failed to build it">
+                    <i class="fa fa-fw fa-warning"></i> {{ release.version }}
+                </a>
```
Comment (jyn514) [inline templates/macros.html:106]: The comments here do not match the code.
```suggestion
            {# If the release is yanked but built, display an warning #}
            {%- if release.yanked and release.build_status -%}
                <a href="{{ release_url }}" class="pure-menu-link warn" title="{{ release_name }} is yanked">
                    <i class="fa fa-fw fa-warning"></i> {{ release.version }}
                </a>

            {# If the release has been yanked and failed to build, display a warning #}
            {%- elif release.yanked and not release.build_status -%}
                <a href="{{ release_url }}" class="pure-menu-link warn"
                    title="{{ release_name }} is yanked and docs.rs failed to build it">
                    <i class="fa fa-fw fa-warning"></i> {{ release.version }}
                </a>
```

**Pull request merged**